10-JavaScript/js-objects.md

```diff
@@ -0,0 +1,141 @@
+> Ada says meow
+```
+
+#### What is `this`?
+
+You may have noticed the keyword `this`.  In JavaScript `this` refers to the current object like `self` in Ruby classes.   Inside an object's instance method you can access an instance variable with `this.<variable name>`
+
+### Object Templates with Prototype
+
+Manual definition is great for 1-off objects, but what if we want to design a template to build numerous objects from, like we do with classes in Ruby?  JavaScript creates object templates differently than Ruby and other Object Oriented Languages.  
+
+#### Ruby-Style:  Classical OOP
+
+In Ruby we create a template for objects by building a class.  We give the class methods and instance variables and then generate instances of the class.
+
+```ruby
+# cat.rb
+class Cat
+  attr_reader :name, :breed
+  def initialize(name, age)
+    @name = name
+    @age = age
+  end
+
+  def info      # Each cat instance is unique
+    "Meet #{@name} who is a #{@age} years old"
+  end
+
+  def self.speak  # Let's assume all cats speak the same way
+    "meow!"
+  end
+end
+
+puts Cat.speak
+amper = Cat.new "Amper", 9
+puts amper.info
+```
+This pattern of defining a class and then instantiating it is very common in object oriented languages, and languages that use it are sometimes said to have _classical_ objects. JavaScript uses a slightly different technique called _prototypical_ objects.
+
+#### Object Templates in JavaScript
+
+In the above Ruby example we have actions we perform on __instances__ (info) and actions we performed at the __class__ level (speak).  To create a template in JavaScript and generate instances involves 3 important steps.  
+
+1. Define a __constructor__ method which sets up the instance, usually initializing instance variables.
+1. Extend the object's __prototype__ to add pieces common to all instances of this type of object, usually instance methods.
+1. Call the constructor using the `new` keyword to create new instances.
+
+Below is an example of a constructor method.  
+```javascript
+let Cat = function(name, age) {
+  this.name = name;
+  this.age = age;
+};
+
+let amper = new Cat("Amper", 7);
+```
+
+By using `new` JavaScript runs the constructor function `Cat` and returns an object with `name`, `age` attributes.
+
+All objects in JavaScript have an object named `prototype` as an attribute.  The prototype has a collection of instance variables and methods inherited from either Object or the object's constructor.  
+
+We can attach methods by extending the object's prototype.  You can do so by assigning the prototype to an object with the methods defined within, or add methods to the prototype object one-by-one.  
+
+```javascript
+Cat.prototype = {
+  info: function() {
+    return `Meet ${this.name} who is a ${this.age} years old`
+  }
+}
+
+Cat.prototype.purr = function() {
+  console.log("Prrrrr");
+}
+
+let octo = new Cat("Octo", 13);
+console.log(octo.info());
+octo.purr();
+> Meet Octo who is a 13 years old
```

Comment:
Can you move the output to correspond to the method call instead of all being at the bottom?